contracts/experimental/burn_zrc2/test/BurnTokensAnyZRC2.test.ts

```diff
@@ -0,0 +1,507 @@
+      ][useraddress];
+      assert(allowed_burn == 0);
+    }
+  });
+
+  it("Should be possible to burn tokens", async () => {
+    setAccount(0);
+    await zrc2contract.Transfer(useraddress, 100);
+    await burncontract.ChangeBurnCancelBlocks(0);
+
+    setAccount(1);
+    // We'll check that the events fire properly here so as not to have another test
+    // case (and thus speed up tests)
+    {
+      let tx = await burncontract.UpdateBurnAllowance(
+        zrc2contract.address,
+        100
+      );
+      expect(tx).to.eventLogWithParams(
+        "UpdateBurnAllowanceSuccess",
+        { value: zrc2contract.address.toLowerCase(), vname: "token_address" },
+        { value: "100", vname: "token_amount" },
+        { value: useraddress, vname: "updated_by" }
+      );
+    }
+    await zrc2contract.Transfer(burncontract.address!, 100);
+
+    // Burns can be finalised by anyone
+    // @todo when we have ZIL transfers, make this a separate account - can't at the
+    // moment 'cos it won't have enough gas.
+    setAccount(0);
+    {
+      let tx = await burncontract.FinaliseBurn(
+        zrc2contract.address,
+        useraddress
+      );
+      expect(tx).to.eventLogWithParams(
+        "TokenBurnConfirmation",
+        { value: zrc2contract.address.toLowerCase(), vname: "token_address" },
+        { value: "100", vname: "token_amount" },
+        { value: useraddress, vname: "burnt_by" }
+      );
+    }
+
+    const pending_burn = await burncontract.pending_burn();
+    const balances = await zrc2contract.balances();
+
+    assert(
+      pending_burn[useraddress] == null,
+      "The user should have no pending burns"
+    );
+    assert(balances[useraddress] == 0, "The user should have no tokens");
+    assert(
+      balances[burncontract.address!.toLowerCase()] == 100,
+      "The contract should have 100 tokens"
+    );
+    assert(
+      balances[adminaddress] == 99900,
+      "The admin must have all unburned tokens"
+    );
+  });
+
+  it("Should be possible to undo your burn before the timeout expires", async () => {
+    setAccount(0);
+    await zrc2contract.Transfer(useraddress, 100);
+    await burncontract.ChangeBurnCancelBlocks(20);
+
+    setAccount(1);
+    await burncontract.UpdateBurnAllowance(zrc2contract.address, 100);
+    await zrc2contract.Transfer(burncontract.address!, 100);
+    await burncontract.CancelBurn(zrc2contract.address);
+
+    const pending_burn = await burncontract.pending_burn();
+    const balances = await zrc2contract.balances();
+
+    assert(
+      pending_burn[useraddress] == null,
+      "The user should have no pending burns"
+    );
+    assert(balances[useraddress] == 100, "The user should have 100 tokens");
+    assert(
+      balances[burncontract.address!.toLowerCase()] == 0,
+      "The contract should have 100 tokens"
+    );
+    assert(
+      balances[adminaddress] == 99900,
+      "The admin must have all unburned tokens"
+    );
+
```

Comment:
Something about the math here doesn't add up. If the user and contract both have 100 tokens and the admin has 99900, then we have 100100 tokens total.